RNMT-4515 Add support for Android 11

luissilvaos opened the pull request:
### Description
This PR introduces an hotfix to ensure the support for Android 11.

The current version of the plugin is not able to open databases when targeting the SDK 30 and running on devices with Android 11.
The issue was already reported storesafe/cordova-sqlite-storage-dependencies#3 and storesafe#954.
From what I was able to understand, it is related with the permissions changes on Android 11, where the SQLite is unable to open the database file using native C primitive `sqlite3_open_v2`:
- https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLGDatabaseHandle.java#L14
- https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLiteNative.java#L60
- https://github.com/liteglue/Android-sqlite-native-driver/blob/master/native/sqlc.c#L40

Causing the handler to crash with a NullPointerException when throwing an exception:
- https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLiteGlueConnection.java#L12

Summing up, this hotfix will enable the usage of the default Android database provider for Android 11 only, by setting the `oldImpl` to `true`, which is the same behaviour of using the `androidDatabaseProvider: 'system'` setting at the Javascript side.

### Context
Fixes: https://outsystemsrd.atlassian.net/browse/RNMT-4515

cmfsotelo commented on src/android/io/sqlc/SQLitePlugin.java:
```diff
@@ -337,6 +337,11 @@ private static String getDBConnectionName(String dbname, JSONObject options) {
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                Log.v(SQLitePlugin.class.getSimpleName(), "Applying hotfix for Android 11+");
+                this.oldImpl = true;
```
Can we have an explanation on how this affects the resolution? maybe a github issue link or something on the commit message?

usernuno replied:
We have this one, but it doesn't refer the `oldImpl` field.

https://github.com/storesafe/cordova-sqlite-storage-dependencies/issues/3

luissilvaos replied:
The `this.oldImpl = true`  is the same result of using the mentioned workaround `androidDatabaseProvider: 'system'`.

luissilvaos replied:
I will add more context to the commit message.

luissilvaos replied:
Done

cmfsotelo commented on CHANGES.md:
```diff
@@ -1,4 +1,7 @@
 # Changes
+### Unreleased
```
Never saw this pattern...
We use this on forks?

luissilvaos replied:
I think not. But we are updating this changelog.

cmfsotelo reviewed (commented):
Minor Question

usernuno reviewed (approved):
Approving, but I agree with @cmfsotelo. We should have a bit more of context in the PR.

oscsx commented on src/android/io/sqlc/SQLitePlugin.java:
```diff
@@ -337,6 +337,11 @@ private static String getDBConnectionName(String dbname, JSONObject options) {
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                Log.v(SQLitePlugin.class.getSimpleName(), "Applying hotfix for Android 11+");
+                this.oldImpl = true;
```
What does this do? Is it equivalent to use the `androidDatabaseProvider: 'system'` setting? There are some glitches when using that setting: https://github.com/storesafe/cordova-sqlite-storage#android-database-provider

luissilvaos replied:
Yes, it is, but it will only be enable for Android 11. Regarding the database lock issue, it seems to no longer be an issue in the most recent Android versions. On the other hand, the data corruption issue appears to be caused by the use of multiple plugins to access the same database.

oscsx replied:
The app always uses the same plugin, so it should not be a problem.

luissilvaos committed:
fix: apply hotfix to support Android 11

The current version of the plugin is not able to open databases when targeting the SDK 30
and running on devices with Android 11.
The issue was already reported https://github.com/storesafe/cordova-sqlite-storage-dependencies/issues/3
and https://github.com/storesafe/cordova-sqlite-storage/issues/954
From what I was able to understand, it is related with the permissions changes on Android 11,
where the SQLite is unable to open the database file using native C primitive `sqlite3_open_v2`:
https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLGDatabaseHandle.java#L14
https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLiteNative.java#L60
https://github.com/liteglue/Android-sqlite-native-driver/blob/master/native/sqlc.c#L40
causing the handler to crash with a NullPointerException when throwing an exception
https://github.com/liteglue/Android-sqlite-connector/blob/master/src/io/liteglue/SQLiteGlueConnection.java#L12

Summing up, this hotfix will enable the usage of the default Android database provider for Android 11 only,
by setting the `oldImpl` to `true`, which is the same behaviour of using the `androidDatabaseProvider: 'system'`
setting at the javascript side.

References https://outsystemsrd.atlassian.net/browse/RNMT-4515

luissilvaos committed:
docs: update CHANGES.md

References https://outsystemsrd.atlassian.net/browse/RNMT-4515

cmfsotelo commented on src/android/io/sqlc/SQLitePlugin.java:
```diff
@@ -337,6 +337,11 @@ private static String getDBConnectionName(String dbname, JSONObject options) {
             this.q = new LinkedBlockingQueue<DBQuery>();
             this.openCbc = cbc;
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
```
hm, using this as such won't make the plugin fail to compile in older versions of mabs, due to not using the SDK 30, thus not knowing the value of Build.VERSION_CODES.R ?

Just a heads up, that this is not "retro compatible", therefore almost renders the if useless.

luissilvaos replied:
Yeah, that's true, and I wondered if we should protected this with the targetSdk too. However, we already have some changes that are not retro compatible with other MABS versions than MABS 7.

cmfsotelo reviewed (approved):
Approving now, but a new question arose, something i only saw now

usernuno reviewed (approved):
Great work with the context!

luissilvaos commented:
Just an additional note, Android R(11) is using SQLite 3.28.0. This version of the plugin is using SQLite 3.26.0.
